src/Arrays/MPIStateArrays.jl

```diff
@@ -334,149 +334,192 @@ end
-    finish_ghost_send!(Q::MPIStateArray)
-end
+function end_ghost_exchange!(Q::MPIStateArray; dependencies = nothing)
+    if any(r -> r == MPI.REQUEST_NULL, Q.recvreq)
+        error("A ghost exchange must begin before it ends.")
+    end
 
-"""
-    finish_ghost_recv!(Q::MPIStateArray)
+    progress = () -> __yield(Q.mpicomm)
+    wait(CPU(), MultiEvent(dependencies), progress)
+
+    __Isend!(Q)
 
-Complete the receive of data and fill the data array on the device
-"""
-function finish_ghost_recv!(Q::MPIStateArray)
     @tic mpi_recvwait
-    # wait on MPI receives
     MPI.Waitall!(Q.recvreq)
+    fill!(Q.recvreq, MPI.REQUEST_NULL)
     @toc mpi_recvwait
 
     @tic mpi_recvcopy
-    # copy data to state vectors
-    prepare_stage!(Q.recv_buffer)
+    event = prepare_stage!(Q.recv_buffer; progress = progress)
     stage = get_stage(Q.recv_buffer)
-    transferrecvbuf!(Q.data, stage, Q.vmaprecv)
+    event = transferrecvbuf!(
+        Q.data,
+        stage,
+        Q.vmaprecv;
+        dependencies = event,
+        progress = progress,
+    )
     @toc mpi_recvcopy
+
+    return event
 end
 
-"""
-    finish_ghost_send!(Q::MPIStateArray)
+function __Irecv!(Q)
+    nnabr = length(Q.nabrtorank)
+    transfer = get_transfer(Q.recv_buffer)
 
-Waits on the send of data to be complete
-"""
-function finish_ghost_send!(Q::MPIStateArray)
-    @tic mpi_sendwait
-    MPI.Waitall!(Q.sendreq)
-    @toc mpi_sendwait
-end
+    for n in 1:nnabr
+        # If this fails we haven't waited on previous recv!
+        @assert Q.recvreq[n].buffer == nothing
 
-# {{{ MPI Buffer handling
-function fillsendbuf!(sendbuf, buf, vmapsend)
-    if length(vmapsend) > 0
-        Np = size(buf, 1)
-        nvar = size(buf, 2)
-
-        event = Event(device(buf))
-        event = knl_fillsendbuf!(device(buf), 256)(
-            Val(Np),
-            Val(nvar),
-            sendbuf,
-            buf,
-            vmapsend,
-            length(vmapsend);
-            ndrange = length(vmapsend),
-            dependencies = (event,),
+        Q.recvreq[n] = MPI.Irecv!(
+            (@view transfer[:, Q.nabrtovmaprecv[n]]),
+            Q.nabrtorank[n],
+            666,
+            Q.mpicomm,
         )
-        wait(device(buf), event)
     end
 end
 
-function transferrecvbuf!(buf, recvbuf, vmaprecv)
-    if length(vmaprecv) > 0
-        Np = size(buf, 1)
-        nvar = size(buf, 2)
+function __Isend!(Q)
+    nnabr = length(Q.nabrtorank)
+    transfer = get_transfer(Q.send_buffer)
 
-        event = Event(device(buf))
-        event = knl_transferrecvbuf!(device(buf), 256)(
-            Val(Np),
-            Val(nvar),
-            buf,
-            recvbuf,
-            vmaprecv,
-            length(vmaprecv);
-            ndrange = length(vmaprecv),
-            dependencies = (event,),
+    for n in 1:nnabr
+        Q.sendreq[n] = MPI.Isend(
+            (@view transfer[:, Q.nabrtovmapsend[n]]),
+            Q.nabrtorank[n],
+            666,
```

Comment:
Yes, you still need to pass a tag to MPI Isend and Irecv.  We just don't need a different one for each MPIStateArray since they each have their own dup'd communicator.

Reply:
would it be better to use `MPI.MPI_ANY_TAG`?

Reply:
No since you still want to make sure that we get the update that is meant for the IRecv/ISend pair, while that should be the only traffic on the communicator (since it is now split), someone might add more ISend/IRecv later on.